orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -308,7 +308,30 @@ Status GradientGraphBuilder::Build(const std::unordered_set<std::string>* p_init
     }
   }
 
-  return GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs, p_initializer_names_to_preserve);
+  ORT_RETURN_IF_ERROR(GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs, p_initializer_names_to_preserve));
+
+  // For ORTModule, the graph will be sent back to frontend for cache. Frontend will use inference session to run the
+  // cached graph. We need to save the new NodeArgs to value_info so we will not loss those shape/type information
+  // during graph to graph_proto conversion. Skip the graph outputs as they are already in the output section in
```

Comment:
From the AddValueInfo impl, duplicated insert will error out: "Error: trying to add an existing value info." 

FYI, AddValueInfo has a really bad impl, where it iterate through the value_info to check duplicate. 
We should use std::unordered_set for value_info_

Reply:
with populated value_info, we should start seeing shape information in saved *_training.onnx and *_optimized_training.onnx graph. You can please check to verify.

Reply:
Yes, I can see these information in the graph after the fix.